### ISSUE TITLE
mountain car doesn't work with n_envs > 1

### ISSUE THREAD

**Issue by jlack1987** [adepierre/torchRL#1]
Not sure if this is maintained or not, but i've been using it to help me learn and I noticed that for some reason the mountain car example doesn't work for n_envs > 1. The Pendulum example does work though with multiple environments, and since they share most of the code I'm not sure where the discrepancy is. Going to keep digging and see what I can find

**Comment by adepierre** [comment]
Hello, yeah I don't think it's "maintained" per se as it was just a "here is my code if you want" not a real project I continued working on. But I could take a look if I can find some time.

Could you detail what doesn't work? Like does it crash or just it's not learning properly? Cause if I remember correctly, the mountain car env is like very sensitive to parameters and quite hard to train, so I wouldn't be surprised if changing the number of parallel envs would also require tweaking other parameters to make it learn properly.
I think I just copied the parameters from stable baselines at the time so the comparison was somewhat fair and useful.

**Comment by jlack1987** [comment]
Yeah it seems with any number of environments more than 1 the result is that the resulting policy doesn't put out much effort.  With 1 environment it produces great behavior, but with any more than that it produces the same bad behavior. I could be wrong, but my thought was given the same hyperparameters(except n_envs which i'm increasing) I would expect similar performance from training as long as enough steps are run. That may not be a great assumption though

**Comment by jlack1987** [comment]
ok well I think I may have figured it out. It looks like the reward structure for mountain car is just minimization of effort plus a big reward if it hits the goal. Could be cases during training especially early on if it doesn't find success that it doesn't know about that terminal reward and just shoots to minimize effort which results in the behavior I was seeing where it just doesn't try. I added a term to the reward to reward it for being closer to the goal and in that case we get good behavior even for multiple env's. I can make a PR if you'd like so future ppl don't run into the same problem while trying to learn from your code which is a fantastic learning tool 😄 🙏

**Comment by adepierre** [comment]
Yeah this is why this is considered a "hard" problem/env in RL. Changing the reward makes the problem simpler to solve, but also breaks the consistency with the corresponding gym environment, so not sure this is actually a good idea.

I guess this environment is on purpose made "hard" to test different algorithms on envs with a very sparse reward. Obviously on real usecases you would tweak the reward so you can learn more easily, but for a benchmark "standard" env, it's usually better to keep everything but the algorithm constant.

I'm glad you find my old code useful 🙂

**Comment by jlack1987** [comment]
Ah yep you are correct, without the additional term w/just the sparse reward it's doable but definitely difficult. In that case I'll close this and thanks once more for letting us learn from your code!
